app/components/UserModal.tsx

```diff
@@ -0,0 +1,180 @@
+import { useState } from "react";
+import Alert from "react-bootstrap/Alert";
+import Button from "react-bootstrap/Button";
+import Form from "react-bootstrap/Form";
+import Modal from "react-bootstrap/Modal";
+import Spinner from "react-bootstrap/Spinner";
+import { useForm, SubmitHandler, Controller } from "react-hook-form";
+import { User, UserAPIError } from "@/types/users";
+import { useToken } from "@/utils/auth";
+
+interface UserModalProps {
+  onClose: () => void;
+  onSubmitCallback: (user: User) => Promise<void>;
+  show: boolean;
+  user?: User;
+}
+
+type UserInputs = {
+  name: string;
+  email: string;
+  app_metadata: { roles: string[] };
+};
+
+const USER_DEFAULTS = {
+  name: undefined,
+  email: undefined,
+  app_metadata: { roles: ["readonly"] },
+};
+
+export default function UserModal({
+  onClose,
+  onSubmitCallback,
+  show,
+  user,
+}: UserModalProps) {
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const token = useToken();
+  const {
+    register,
+    handleSubmit,
+    control,
+    formState: { errors, isDirty },
+  } = useForm<UserInputs>({
```

Comment:
react-hook-form with typescript is super easy—just pass in the type and suddenly the whole form is typesafe 🔥